File: app/src/main/java/eu/kanade/tachiyomi/data/sync/WebDavTransport.kt
package eu.kanade.tachiyomi.data.sync

import eu.kanade.tachiyomi.network.NetworkHelper
import eu.kanade.tachiyomi.network.await
import okhttp3.Credentials
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import java.io.File

/**
 * Stores a single .tachibk file on a WebDAV server. The merge happens client-side via the
 * BackupRestorer's per-row timestamp-based logic.
 *
 * Auth: HTTP Basic. The username comes from a dedicated preference; the password is reused from
 * the existing sync apiKey field.
 */
class WebDavTransport(
    private val host: String,
    private val username: String,
    private val password: String,
    private val networkHelper: NetworkHelper,
) : SyncTransport.FileStorage {

    private val fileUrl: String
        get() = "$host/$FILE_NAME"

    private val authHeader: String
        get() = Credentials.basic(username, password)

    override suspend fun pull(): ByteArray? {
        val request = Request.Builder()
            .url(fileUrl)
            .header("Authorization", authHeader)
            .get()
            .build()

        val response = networkHelper.client.newCall(request).await()
        try {
            return when (response.code) {
                in 200..299 -> response.body.bytes().takeIf { it.isNotEmpty() }
                404 -> null
                401, 403 -> throw SyncException("WebDAV auth failed (HTTP ${response.code})")
                else -> throw SyncException("WebDAV pull failed (HTTP ${response.code})")
            }
        } finally {
            response.close()
        }
    }

    override suspend fun push(payload: File) {
        val request = Request.Builder()
            .url(fileUrl)
            .header("Authorization", authHeader)
            .put(payload.asRequestBody(OCTET_STREAM))
            .build()

        val response = networkHelper.client.newCall(request).await()
        try {
            if (!response.isSuccessful) {
                throw SyncException("WebDAV push failed (HTTP ${response.code})")
            }
        } finally {
            response.close()
        }
    }

    companion object {
        private const val FILE_NAME = "mihon-sync.tachibk"
        private val OCTET_STREAM = "application/octet-stream".toMediaTypeOrNull()
    }
}

File: flutter/lib/data/sync/webdav_transport.dart
/// WebDAV transport — single .tachibk file at a fixed path. HTTP Basic
/// auth. The merge happens client-side via [BackupRestorer]'s per-row
/// `lastModifiedAt` logic.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:dio/dio.dart';

import 'sync_transport.dart';

class WebDavTransport extends SyncTransportFileStorage {
  WebDavTransport({
    required String host,
    required this.username,
    required this.password,
    required this.dio,
  }) : host = host.trimRight().replaceAll(RegExp(r'/+$'), '');

  final String host;
  final String username;
  final String password;
  final Dio dio;

  /// Same filename Mihon uses so a Mohyeong install can drop into the
  /// same WebDAV folder as an existing Mihon install with no migration.
  static const String _fileName = 'mohyeong-sync.tachibk';

  String get _fileUrl => '$host/$_fileName';

  String get _authHeader =>
      'Basic ${base64Encode(utf8.encode('$username:$password'))}';

  @override
  Future<Uint8List?> pull() async {
    try {
      final response = await dio.get<List<int>>(
        _fileUrl,
        options: Options(
          responseType: ResponseType.bytes,
          headers: {'Authorization': _authHeader},
          // We want to inspect the status code rather than throw on 404.
          validateStatus: (s) => s != null && (s == 404 || (s >= 200 && s < 300)),
        ),
      );
      if (response.statusCode == 404) return null;
      final bytes = response.data;
      if (bytes == null || bytes.isEmpty) return null;
      return Uint8List.fromList(bytes);
    } on DioException catch (e) {
      final code = e.response?.statusCode;
      if (code == 401 || code == 403) {
        throw SyncException('WebDAV auth failed (HTTP $code)');
      }
      throw SyncException('WebDAV pull failed (HTTP ${code ?? '?'})');
    }
  }

  @override
  Future<void> push(Uint8List payload) async {
    try {
      await dio.put<void>(
        _fileUrl,
        data: Stream.value(payload),
        options: Options(
          headers: {
            'Authorization': _authHeader,
            'Content-Type': 'application/octet-stream',
            'Content-Length': payload.length,
          },
          validateStatus: (s) => s != null && s >= 200 && s < 300,
        ),
      );
    } on DioException catch (e) {
      throw SyncException(
          'WebDAV push failed (HTTP ${e.response?.statusCode ?? '?'})');
    }
  }
}
